Add bezier preset and refactor bezier editor

TokageItLab opened the pull request:
Co-authored-by: Razoric480 <razoric480@gmail.com>

Supersedes/Salvaged #57920.

![image](https://user-images.githubusercontent.com/61938263/185730620-ac326ae8-4698-4efa-9a9f-a02a8b5a1290.png) ![image](https://user-images.githubusercontent.com/61938263/185730630-43ebdcb2-0993-4520-882b-ecd7b29cbb8b.png)


Add a Bezier preset. Also corrected incorrect signals (#64097). `handle_mode` is now only a valid property if `TOOLS = true`.

The undo-redo is not working, but it is broken before this PR, so it has nothing to do with this PR itself and will need to be fixed later.

Fixes https://github.com/godotengine/godot-proposals/issues/3561. Fixes #64097.

Razoric480 commented:
@TokageItLab 
Code looks great and solves the issues that were raised in the previous PR. Thanks for picking up my slack :) It's appreciated.

Compiled and gave it a spin. A couple apparent bugs:
- Using right click to insert a new key does not cause a redraw, until you do something that does refresh. Only then does the new keyframe appears.
- Focus with no keyframe selected seems to break <kbd>F</kbd>ocusing, zooming out to minimum and having an error state vertically. I suspect changes to animation_bezier_editor.cpp's focus logic around the 875-901 lines. Not sure if the bug was there before. If we have no keyframes selected, maybe we should just focus on the length of the animation horizontally and the height of the top keyframes vertically.
- The changes from #55030 seem to be gone, as I can no longer select multiple bezier tracks' keys simultaneously. Not sure if this happened here, or if something else happened since that PR and this one.

Other than those three things, it works pretty well.

TokageItLab commented:
@Razoric480 Thank you for testing, I'll take a look later.

TokageItLab commented:
> Using right click to insert a new key does not cause a redraw, until you do something that does refresh. Only then does the new keyframe appears.

Fixed.

> Since https://github.com/godotengine/godot/pull/55030 seem to have gotten partly reverted. All curves draw simultaneously, but they cannot be selected simultaneously. Not sure if this happened here, or if something else happened since that PR and this one.

It seems to be a regression that occurred in #57920. Fixed.

> Focus with no keyframe selected seems to break Focusing, zooming out to minimum and having an error state vertically. I suspect changes to animation_bezier_editor.cpp's focus logic around the 875-901 lines. Not sure if the bug was there before. If we have no keyframes selected, maybe we should just focus on the length of the animation horizontally and the height of the top keyframes vertically.

It seems to be a bug that occurred in master before this PR. I don't think it is critical for now, so I think we can fix it in Beta or later if you need.

Razoric480 reviewed (approved):
Thanks! Works great now!

And yeah, we can fix and improve the focus key in a subsequent PR.

YuriSizov commented on doc/classes/Animation.xml:
```diff
@@ -188,7 +187,7 @@
 			<param index="0" name="track_idx" type="int" />
 			<param index="1" name="key_idx" type="int" />
 			<param index="2" name="key_handle_mode" type="int" enum="Animation.HandleMode" />
-			<param index="3" name="balanced_value_time_ratio" type="float" default="1.0" />
+			<param index="3" name="key_handle_set_mode" type="int" enum="Animation.HandleSetMode" default="0" />
```
The docs could explain the difference between handle mode and handle set mode.

YuriSizov commented:
I have to say, the icons are very hard to read. Especially the difference between free and balanced. The key part to distinguish them is the handles, and they are the least visible. Can we do something about it?

Razoric480 commented:
@YuriSizov 
How about these?
![BezierHandlesBalanced](https://user-images.githubusercontent.com/2823241/185752500-048e294f-ce66-41e1-985e-0c7e6836e0aa.svg)
![BezierHandlesFree](https://user-images.githubusercontent.com/2823241/185752459-57210acf-7c7c-4731-9660-1a11f4c5fa3e.svg)
![BezierHandlesLinear](https://user-images.githubusercontent.com/2823241/185752460-32380d32-29ee-4c3d-8bb3-405ad3f705c2.svg)
![BezierHandlesMirror](https://user-images.githubusercontent.com/2823241/185752461-8e1bb211-85d2-4f76-94f3-825839959d2d.svg)

YuriSizov commented:
@Razoric480 Lovely! They are still very tiny, which may hinder the experience for some, but it's not something we can fix per icon and more of an overall issue. Definitely clearer this way!

Razoric480 commented:
@TokageItLab
[BezierIcons.zip](https://github.com/godotengine/godot/files/9387374/BezierIcons.zip)
Don't want to force-push on your branch in order to avoid conflicts, so I put them in there. Unless you'd prefer I just make a commit and you can squash it after the fact?

TokageItLab commented:
Changed icons and updated docs.

TokageItLab committed:
Add bezier preset and refactor bezier editor

Co-authored-by: Razoric480 <razoric480@gmail.com>

akien-mga commented on scene/resources/animation.cpp:
```diff
@@ -3799,8 +3905,10 @@ void Animation::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("audio_track_get_key_end_offset", "track_idx", "key_idx"), &Animation::audio_track_get_key_end_offset);
 
-	ClassDB::bind_method(D_METHOD("bezier_track_set_key_handle_mode", "track_idx", "key_idx", "key_handle_mode", "balanced_value_time_ratio"), &Animation::bezier_track_set_key_handle_mode, DEFVAL(1.0));
+#ifdef TOOLS_ENABLED
+	ClassDB::bind_method(D_METHOD("bezier_track_set_key_handle_mode", "track_idx", "key_idx", "key_handle_mode", "key_handle_set_mode"), &Animation::bezier_track_set_key_handle_mode, DEFVAL(HANDLE_SET_MODE_NONE));
 	ClassDB::bind_method(D_METHOD("bezier_track_get_key_handle_mode", "track_idx", "key_idx"), &Animation::bezier_track_get_key_handle_mode);
+#endif // TOOLS_ENABLED
```
We shouldn't bind methods/constants only in `TOOLS_ENABLED` builds. This makes them part of the Core API but only when you run an editor build, which means you could use them while developing your project but it will fail in an exported project. They also appear in the documentation (`--doctool` is a `TOOLS_ENABLED` build), but they shouldn't if they're only meant for editor internal code.

Having `TOOLS_ENABLED` conditional method bindings can also break the Mono bindings in 3.x, I don't know what's the state in 4.0 with the dotnet6 port but either way, this needs a different approach.

TokageItLab replied:
That should only be necessary for UndoRedo, so I'll try to figure out if it is possible to set it on the Editor side.

TokageItLab replied:
That method is now bound to the AnimationTrackEditor side.

TokageItLab commented:
Fixed problems around Undo/Redo. The main problem seemed to be with redrawing timing.

akien-mga commented on scene/resources/animation.cpp:
```diff
@@ -3849,8 +3951,16 @@ void Animation::_bind_methods() {
 	BIND_ENUM_CONSTANT(LOOP_LINEAR);
 	BIND_ENUM_CONSTANT(LOOP_PINGPONG);
 
+#ifdef TOOLS_ENABLED
 	BIND_ENUM_CONSTANT(HANDLE_MODE_FREE);
+	BIND_ENUM_CONSTANT(HANDLE_MODE_LINEAR);
 	BIND_ENUM_CONSTANT(HANDLE_MODE_BALANCED);
+	BIND_ENUM_CONSTANT(HANDLE_MODE_MIRRORED);
+
+	BIND_ENUM_CONSTANT(HANDLE_SET_MODE_NONE);
+	BIND_ENUM_CONSTANT(HANDLE_SET_MODE_RESET);
+	BIND_ENUM_CONSTANT(HANDLE_SET_MODE_AUTO);
+#endif // TOOLS_ENABLED
```
This is the same problem as for the methods, there shouldn't be `TOOLS_ENABLED` only API exposed to bindings. These enums values are not useful to users so they shouldn't be part of the public API.

TokageItLab replied:
Removed these.

akien-mga commented on scene/resources/animation.h:
```diff
@@ -485,7 +498,10 @@ class Animation : public Resource {
 VARIANT_ENUM_CAST(Animation::UpdateMode);
-VARIANT_ENUM_CAST(Animation::HandleMode);
 VARIANT_ENUM_CAST(Animation::LoopMode);
+#ifdef TOOLS_ENABLED
+VARIANT_ENUM_CAST(Animation::HandleMode);
+VARIANT_ENUM_CAST(Animation::HandleSetMode);
+#endif // TOOLS_ENABLED
```
These can likely be removed too I think.

TokageItLab replied:
This is necessary if Enum is used as a function argument. It can be removed if you are using an int as a function argument.

TokageItLab replied:
Well, I recommend leaving this as is. If you must remove this, please let me know again.

akien-mga replied:
Sounds good :)

akien-mga commented:
Thanks!
